Modules/_sre.c

```diff
@@ -563,17 +563,36 @@ pattern_error(Py_ssize_t status)
     }
 }
 
+static int
+pattern_traverse(PatternObject *self, visitproc visit, void *arg)
+{
+    Py_VISIT(self->groupindex);
+    Py_VISIT(self->indexgroup);
+    Py_VISIT(self->pattern);
+    Py_VISIT(Py_TYPE(self));
+    return 0;
+}
+
+static int
+pattern_clear(PatternObject *self)
+{
+    if (self->weakreflist != NULL) {
+        PyObject_ClearWeakRefs((PyObject *) self);
+    }
+    Py_CLEAR(self->groupindex);
+    Py_CLEAR(self->indexgroup);
+    Py_CLEAR(self->pattern);
+    return 0;
+}
+
 static void
 pattern_dealloc(PatternObject* self)
 {
     PyTypeObject *tp = Py_TYPE(self);
 
-    if (self->weakreflist != NULL)
-        PyObject_ClearWeakRefs((PyObject *) self);
-    Py_XDECREF(self->pattern);
-    Py_XDECREF(self->groupindex);
-    Py_XDECREF(self->indexgroup);
-    PyObject_Free(self);
+    PyObject_GC_UnTrack(self);
+    (void)pattern_clear(self);
+    PyObject_GC_Del(self);
```

Comment:
I would prefer to use `tp->tp_free(self);` here. Same remark for other dealloc functions.